### PR TITLE
Ensure oversized data frames are not written to spdystreams

### LINKED ISSUE #460: Copy of 16M+ file to container causes invalid data frame error
Hello,

We have been trying to implement a copy mechanism in to a container.
We are using this code:

```
stdin := bytes.NewReader(byteArray)
req := clientset.Core().RESTClient().Post().
	Resource("pods").
	Name("podName").
	Namespace("namespace").
	SubResource("exec")
scheme := runtime.NewScheme()
check(err)

parameterCodec := runtime.NewParameterCodec(scheme)
req.VersionedParams(&core_v1.PodExecOptions{
	Command:   strings.Fields("cp /dev/stdin /tmp/file"),
	Container: containerName,
	Stdin:     stdin != nil,
	Stdout:    true,
	Stderr:    true,
	TTY:       false,
}, parameterCodec)

exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
check(err)

var stdout, stderr bytes.Buffer
err = exec.Stream(remotecommand.StreamOptions{
	Stdin:  stdin,
	Stdout: &stdout,
	Stderr: &stderr,
	Tty:    false,
})
check(err)
```

The complete code can be found [here](https://github.com/maorfr/skbn/blob/master/pkg/skbn/kube.go)

We noticed that when trying to copy a file which is 16M and up, the action is not successful and we get an output of `E0904 12:56:33.953212   15894 v2.go:105] invalid data frame`.

Any file that is under 16M (16777215 and less) is being copied without any error.

Can anyone help on this?

Thanks in advance!

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/client-go/issues/460

If the Stdin provided as input [implements a `WriteTo` function](https://golang.org/pkg/bytes/#Buffer.WriteTo), io.Copy uses it to [attempt to write the entire buffer to the stream in one call](https://golang.org/pkg/io/#Copy), which [results in a data frame that is too large](https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/docker/spdystream/stream.go#L66-L73) and silently drops the outgoing frame.

**Does this PR introduce a user-facing change?**:
```release-note
client-go: fixes sending oversized data frames to spdystreams in remotecommand.NewSPDYExecutor
```

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-api-machinery-pr-reviews 
/cc @deads2k @caesarxuchao

**Comment by deads2k** [comment]
I think I get it now.

/lgtm

File: staging/src/k8s.io/client-go/tools/remotecommand/reader.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"io"
+)
+
+// readerWrapper delegates to an io.Reader so that only the io.Reader interface is implemented.
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/remotecommand/reader.go]: avoiding passing in a reader that implements WriteTo (linked to the problematic code path in the description)

[the remark this answers is not in the document]
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/remotecommand/reader.go]: Ah I didn't read the PR description carefully enough. But future readers of this code won't see it at all, so can you repeat that in the comment here?

[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/remotecommand/reader.go]: Sure

**Comment by liggitt** [comment]
/hold

will address comments

**Comment by liggitt** [comment]
/hold cancel

added comments and deflaked

**Comment by liggitt** [comment]
only thing changed was the comment and the unit test timeout, retagging

**Comment by lavalamp** [comment]
/lgtm

Isn't this a bug in the spdystream library? Shouldn't it split up the data?

**Comment by liggitt** [comment]
> Isn't this a bug in the spdystream library? Shouldn't it split up the data?

I think it's working as intended... they document their impl of Write to mean 1 write == 1 frame. That said, that doesn't play nice with standard io libraries

**Pull request merged**